agent_build/package_builders.py

```diff
@@ -0,0 +1,802 @@
+    """
+    # Compile the strings into actual regular expression match objects.
+    matchers = []
+    for file_name in file_names:
+        matchers.append(re.compile(str(file_name)))
+
+    # Walk down the current directory.
+    for root, dirs, files in os.walk(dir_path.absolute()):
+        # See if any of the files at this level match any of the matchers.
+        for file_path in files:
+            for matcher in matchers:
+                if matcher.match(file_path):
+                    # Delete it if it did match.
+                    os.unlink(os.path.join(root, file_path))
+                    break
+
+
+def recursively_delete_dirs_by_name(root_dir: Union[str, pl.Path], *dir_names: str):
+    """
+    Deletes any directories that are in the current working directory or any of its children whose file names
+    match the specified regular expressions.
+
+    This will recursively examine all children of the current working directory.
+
+    If a directory is found that needs to be deleted, all of it and its children are deleted.
+
+    @param dir_names: A variable number of strings containing regular expressions that should match the file names of
+        the directories that should be deleted.
+    """
+
+    # Compile the strings into actual regular expression match objects.
+    matchers = []
+    for dir_name in dir_names:
+        matchers.append(re.compile(dir_name))
+
+    # Walk down the file tree, top down, allowing us to prune directories as we go.
+    for root, dirs, files in os.walk(root_dir):
+        # Examine all directories at this level, see if any get a match
+        for dir_path in dirs:
+            for matcher in matchers:
+                if matcher.match(dir_path):
+                    shutil.rmtree(os.path.join(root, dir_path))
+                    # Also, remove it from dirs so that we don't try to walk down it.
+                    dirs.remove(dir_path)
+                    break
+
+
+# Special global collection of all builders. It can be used by CI/CD scripts to find needed package builder.
+ALL_PACKAGE_BUILDERS: Dict[str, "PackageBuilder"] = {}
+
+
+class PackageBuilder(abc.ABC):
+    """
+        Base abstraction for all Scalyr agent package builders. it can perform build of the package directly on the
+    current system or inside docker.
+        It also uses ':py:module:`agent_tools.environment_deployments` features to define and deploy its build
+        environment in order to be able to perform the actual build.
+    """
+
+    # Type of the package to build.
+    PACKAGE_TYPE: constants.PackageType
+
+    # Add agent source code as a bundled frozen binary if True, or
+    # add the source code as it is.
+    USE_FROZEN_BINARIES: bool = True
+
+    # Specify the name of the frozen binary, if it is used.
+    FROZEN_BINARY_FILE_NAME = "scalyr-agent-2"
+
+    # The type of the installation. For more info, see the 'InstallType' in the scalyr_agent/__scalyr__.py
+    INSTALL_TYPE: str
```

Comment:
It actually seems that ``*_INSTALL`` constants in that module are ints and not strs?

In any case, once we drop support for Python < 3.6 we should probably just update the code utilize enum type for that.

Reply:
Yeah, I just don't want to pull the scalyr_agent package because it will depend on a bunch of third-party libraries and I just want to keep build libraries as independent as possible, just in case.